mrmustard/lab_dev/circuit_components.py

```diff
@@ -470,11 +530,37 @@ def __matmul__(self, other: CircuitComponent | Scalar) -> CircuitComponent:
         Contracts ``self`` and ``other`` without adding adjoints.
         It allows for a more custom way of contracting components.
         """
+        try:
+            return other._rrshift_(self)
+        except AttributeError:
+            pass
         if isinstance(other, (numbers.Number, np.ndarray)):
             return self * other
+
         wires_result, perm = self.wires @ other.wires
         idx_z, idx_zconj = self._matmul_indices(other)
-        rep = self.representation[idx_z] @ other.representation[idx_zconj]
+
+        if isinstance(self.representation, Bargmann) and isinstance(other.representation, Bargmann):
+            rep = self.representation[idx_z] @ other.representation[idx_zconj]
+            rep = rep.reorder(perm) if perm else rep
+            return CircuitComponent._from_attributes(rep, wires_result, None)
```

Comment:
Yep, this is an excellent question. We do, and in fact `__rshift__` is overrided in `Ket`, `DM`, `Unitary` and `Channel` because in those cases we do know more and we can create more accurate objects. In full generality (such as the generality that `__matmul__` grants us) nothing can be said.